src/hir-lowering.md

```diff
@@ -1 +1,119 @@
 # HIR lowering
+
+The HIR -- "High-level IR" -- is the primary IR used in most of
+rustc. It is a desugared version of the "abstract syntax tree" (AST)
+that is generated after parsing, macro expansion, and name resolution
+have completed. Many parts of HIR resemble Rust surface syntax quite
+closely, with the exception that some of Rust's expression forms have
+been desugared away (as an example, `for` loops are converted into a
+`loop` and do not appear in the HIR).
+
+This chapter covers the main concepts of the HIR.
+
+### Out-of-band storage and the `Crate` type
+
+The top-level data-structure in the HIR is the `Crate`, which stores
+the contents of the crate currently being compiled (we only ever
+construct HIR for the current crate). Whereas in the AST the crate
+data structure basically just contains the root module, the HIR
+`Crate` structure contains a number of maps and other things that
+serve to organize the content of the crate for easier access.
+
+For example, the contents of individual items (e.g., modules,
+functions, traits, impls, etc) in the HIR are not immediately
+accessible in the parents. So, for example, if had a module item `foo`
+containing a function `bar()`:
+
+```
+mod foo {
+  fn bar() { }
+}
+```
```

Comment:
I wonder if we can start testing these code samples. That might .. actually be a way for us to test links too? Gonna take a bit of work to make that easy I suppose.

Reply:
A potentially useful starting point: https://github.com/budziq/rust-skeptic/blob/master/README.md

Reply:
Now version `0.1.0` of `mdbook`, we might be able to use my [mdbook-test] backend. It's essentially gets a copy of the document from `mdbook` and then runs `rust-skeptic` over it. 

I found it worked pretty well locally but the CI tests were quite flaky and would fail for unknown reasons. It also needs to be updated to pull `mdbook` from crates.io instead of the git repository.

[mdbook-test]: https://github.com/Michael-F-Bryan/mdbook-test